trustchain-ion/src/root.rs

```diff
@@ -0,0 +1,168 @@
+use chrono::NaiveDate;
+use futures::{future, StreamExt};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use trustchain_core::utils::get_did_from_suffix;
+
+use crate::{
+    utils::{
+        block_height_range_on_date, locate_transaction, query_mongodb_on_interval, transaction,
+    },
+    TrustchainBitcoinError, TrustchainMongodbError, ION_TEST_METHOD, MONGO_FILTER_DID_SUFFIX,
+    MONGO_FILTER_TXN_TIME,
+};
+
+/// An error relating to the root DID.
+#[derive(Error, Debug)]
+pub enum TrustchainRootError {
+    /// Bitcoin RPC interface error while processing root event date.
+    #[error("Bitcoin RPC error while processing root event date.")]
+    BitcoinRpcError(TrustchainBitcoinError),
+    /// Mongo DB error while processing root event date.
+    #[error("Mongo DB error while processing root event date.")]
+    MongoDbError(TrustchainMongodbError),
+    /// Failed to identify unique root DID.
+    #[error("No unique root DID on date: {0}")]
+    NoUniqueRootEvent(NaiveDate),
+    /// Invalid date.
+    #[error("Invalid date: {0}-{1}-{2}")]
+    InvalidDate(i32, u32, u32),
+    /// Failed to parse block height.
+    #[error("Failed to parse block height: {0}")]
+    FailedToParseBlockHeight(String),
+}
+
+impl From<TrustchainBitcoinError> for TrustchainRootError {
+    fn from(err: TrustchainBitcoinError) -> Self {
+        TrustchainRootError::BitcoinRpcError(err)
+    }
+}
+
+impl From<TrustchainMongodbError> for TrustchainRootError {
+    fn from(err: TrustchainMongodbError) -> Self {
+        TrustchainRootError::MongoDbError(err)
+    }
+}
+
+/// Struct representing a root DID candidate.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct RootCandidate {
+    pub did: String,
+    pub txid: String,
+    pub block_height: u64,
+}
+
+/// Identifies potential root DIDs whose (UTC) timestamp matches a given date.
+/// Root DID candidates are those that are found in ION create operations with
+/// operation index zero (opIndex = 0). As such, root DIDs must be created in
+/// the first DID operation associated with a particular Bitcoin transaction.
+pub async fn root_did_candidates(
+    date: NaiveDate,
+) -> Result<Vec<RootCandidate>, TrustchainRootError> {
+    let block_height_range = block_height_range_on_date(date, None, None)?;
+    let cursor =
+        query_mongodb_on_interval(block_height_range.0 as u32, block_height_range.1 as u32).await?;
+
+    // The mongodb Cursor instance streams all bson documents that:
```

Comment:
Remove?

Reply:
Will leave for now as later refactoring will make obsolete.